mplc/multi_partner_learning.py

```diff
@@ -663,6 +664,577 @@ def fit_minibatch(self):
+            store_model_weights(model, model_stateholder)
+            for p_id, minibatch in enumerate(partners_minibatches):  # minibatch == (x,y)
+                x_minibatch = tf.unstack(minibatch[0], axis=0)
+                y_minibatch = tf.unstack(minibatch[1], axis=0)
+                for x, y in zip(x_minibatch, y_minibatch):
+                    with tf.GradientTape() as tape:
+                        y_pred = model(x)
+                        loss = model.compiled_loss(y, y_pred)
+                    model.compiled_metrics.update_state(y, y_pred)
+                    partners_optimizers[p_id].minimize(loss, model.trainable_weights, tape=tape)  # local optimization
+                for grad_per_layer, w_old, w_new in zip(partners_grads[p_id], model_stateholder,
+                                                        model.trainable_weights):
+                    grad_per_layer.assign(
+                        (w_old - w_new))  # get the gradient as theta_before_minibatch - theta_after_minibatch
+                load_stored_weights(model, model_stateholder)
+
+            global_grad = aggregeted_grads(tuple(partners_grads),
+                                           agg_w)  # at the end of the minibatch, aggregate the gradients.
+
+            # for each local optimizer we perform an optimization step with the global gradient,
+            # to update the momentums and learning rate with this global gradient.
+            # We informe the local optimizer of the global optimization.
+
+            for optimizer in partners_optimizers:
+                optimizer.apply_gradients(zip(global_grad, model.trainable_weights))
+                # in fact the optimization of the weights must be done by the global optimizer,
+                # so we reset the weights after the optimization
+                load_stored_weights(model, model_stateholder)
+
+            model.optimizer.apply_gradients(
+                zip(global_grad, model.trainable_weights))  # apply the global gradients with the global optimizer
+
+        self.timer = time.time()
+        for e in range(self.epoch_count):
+            self.epoch_timer = time.time()
+            for partners_minibatches in zip(*self.train_dataset):  # <- partners_minibatches == [(x, y)] * nb_partners
+                fit_minibatch(self.model, self.model_stateholder, partners_minibatches, self.partners_optimizers,
+                              self.partners_grads, self.agg_w)
+            epoch_history = self.get_epoch_history()  # compute val and train acc and loss.
+            # add the epoch history to self history, and log epoch number, and metrics values.
+            self.log_epoch(e, epoch_history)
+
+            self.epochs_index += 1
+            if self.early_stop():
+                break
+
+        self.log_end_training()
+
+
 # Supported multi-partner learning approaches
 
 MULTI_PARTNER_LEARNING_APPROACHES = {
```

Comment:
Shouldn't you add them there?

Reply:
I am not sure than everything will run ok (I only tested these methods in separated notebook) so I have some tests to do before adding them. But I will have to add them here yes